### PR TITLE
Watch mesh configuration ConfigMap

### PR DESCRIPTION
**Pull request by tbarrella**
As part of #27707

[ ] Configuration Infrastructure
[ ] Docs
[ ] Installation
[X] Networking
[ ] Performance and Scalability
[ ] Policies and Telemetry
[ ] Security
[ ] Test and Release
[ ] User Experience
[ ] Developer Infrastructure

### CODE REVIEW

**Review by costinm — APPROVED** [review]
Looks good, we may need to refine it but seems safe.

File: pilot/pkg/bootstrap/server.go [hunk 1]
```diff
@@ -218,14 +218,15 @@ func NewServer(args *PilotArgs) (*Server, error) {
...
 
-	// TODO: revert to watching k8s (and merge with the file)
 	s.initMeshConfiguration(args, s.fileWatcher)
 
 	// Apply the arguments to the configuration.
 	if err := s.initKubeClient(args); err != nil {
 		return nil, fmt.Errorf("error initializing kube client: %v", err)
 	}
 
+	s.watchMeshConfiguration(args)
```
Comment (hzxuzhonghu) [inline pilot/pkg/bootstrap/server.go:228]: This should be in initMeshConfiguration and only watch it when local file not exist.

They are independent

Comment (tbarrella) [reply pilot/pkg/bootstrap/server.go:228]: I tried this originally but it didn't work; `watchMeshConfiguration` needs `s.kubeClient` to be set, which is done in `initKubeClient`, and `initKubeClient` needs some mesh config (like the default one) to know whether to initialize `s.kubeClient`. Maybe renaming it to `maybeWatchMeshConfiguration`?

File: pilot/pkg/bootstrap/mesh.go [hunk 2]
```diff
@@ -51,6 +61,19 @@ func (s *Server) initMeshConfiguration(args *PilotArgs, fileWatcher filewatcher.
...
+// maybeWatchMeshConfiguration watches mesh configuration from a ConfigMap if no local file is present.
+// This is done after initMeshConfiguration, which must be done to initialize the Kubernetes client.
+func (s *Server) maybeWatchMeshConfiguration(args *PilotArgs) {
+	if s.kubeClient == nil {
+		return
+	}
+	if _, err := os.Stat(args.MeshConfigFile); os.IsNotExist(err) {
+		configMapName := getConfigMapName(args.Revision)
+		s.environment.Watcher = mesh.NewConfigMapWatcher(
+			s.kubeClient, constants.IstioSystemNamespace, configMapName, configMapKey)
```
Comment (howardjohn) [inline pilot/pkg/bootstrap/mesh.go:73]: we should use the passed in namespace in `args`. `IstioSystemNamespace` is only the default

**Review by howardjohn — COMMENTED** [review]
Seems like currently logic is
```
ReadMeshConfigFile. If not set, use defaults, which has k8s
If mesh config had k8s, create kube client
If mesh config had k8s, replace watch with configmap watch
```

This gets a bit wonky since we have mesh config watchers in multiple places, and even create a watcher then overwrite it. Since the logic is decoupled some of the logic doesn't match either (in the initMeshConfig we check the meshconfigfile is set, but in maybeWatchMeshConfiguration we check it actually exists, for example.

Would it make sense to reshape it like:

```
CheckHasKubeRegistry -> read mesh config file 1x, if it exists and is configured, and check for k8s registry
If had k8s, create kube client
initMeshConfig:
  if hasK8sClient && configmap exists { watch configmap }
  else if file configured { watch file }
  else fixed watcher for defaults
```

to consolidate all watching logic?


I think the code here is correct so not a big issue - but if we don't change it we should have comments around the ordering since its not immediately obvious. will approve if you fix the namespace comment

**Comment by tbarrella** [comment]
Thank you, that is cleaner. I can have `initKubeClient` first and have it call `ReadMeshConfig` instead of `s.environment.Mesh`. Although I'd like to keep the logic of

- If the local file exists, watch it
- Otherwise, watch the ConfigMap

both because I don't want to change the existing behavior by default and also to be more responsive to ConfigMap changes, for example if there isn't one at first and then it's added later

**Review by howardjohn — APPROVED** [review]

**Pull request merged**